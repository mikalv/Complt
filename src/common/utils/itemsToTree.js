function itemsToTree(items, topLevelId) {
  if (items.length === 0) return undefined;
  const topLevelItem = items.find(item => item._id === topLevelId);
  const topLevelItemChildren = topLevelItem.children.map((childId) => {
    const item = items.find(itemInFind => itemInFind._id === childId);
    if (item === undefined) return null;
    if (item.isProject) {
      return itemsToTree(items, item._id);
    }
    return item;
  });
  return {
    ...topLevelItem,
    children: topLevelItemChildren,
  };
}

export default itemsToTree;
