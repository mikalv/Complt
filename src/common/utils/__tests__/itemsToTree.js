import itemsToTree from '../itemsToTree';

const items = [
  { _id: 'item1', isProject: false, isCompleted: true, tags: [] },
  { _id: 'item2', isProject: false, isCompleted: true, tags: [] },
  { _id: 'item3', isProject: false, isCompleted: false, tags: [] },
  { _id: 'item4', isProject: false, isCompleted: true, tags: [] },
  { _id: 'item5', isProject: false, isCompleted: true, tags: [] },
  { _id: 'item6', isProject: false, isCompleted: false, tags: [] },
  { _id: 'item7', isProject: false, isCompleted: true, tags: [] },
  { _id: 'item8', isProject: false, isCompleted: true, tags: [] },
  { _id: 'item9', isProject: false, isCompleted: false, tags: [] },
  {
    _id: 'item11',
    isProject: true,
    children: ['item4', 'item5', 'item6', 'item8'],
  },
  {
    _id: 'item12',
    isProject: true,
    children: ['item3', 'item11', 'item10', 'item7'],
  },
  {
    _id: 'root',
    isProject: true,
    children: ['item1', 'item2', 'item12', 'item9'],
  },
];

describe('itemsToTree', () => {
  it('returns an object with all the children properties being the actual children instead of their ids', () => {
    expect(itemsToTree(items, 'root')).toMatchSnapshot();
  });
});
