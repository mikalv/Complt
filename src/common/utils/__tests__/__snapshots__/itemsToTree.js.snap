exports[`itemsToTree returns an object with all the children properties being the actual children instead of their ids 1`] = `
Object {
  "_id": "root",
  "children": Array [
    Object {
      "_id": "item1",
      "isCompleted": true,
      "isProject": false,
      "tags": Array [],
    },
    Object {
      "_id": "item2",
      "isCompleted": true,
      "isProject": false,
      "tags": Array [],
    },
    Object {
      "_id": "item12",
      "children": Array [
        Object {
          "_id": "item3",
          "isCompleted": false,
          "isProject": false,
          "tags": Array [],
        },
        Object {
          "_id": "item11",
          "children": Array [
            Object {
              "_id": "item4",
              "isCompleted": true,
              "isProject": false,
              "tags": Array [],
            },
            Object {
              "_id": "item5",
              "isCompleted": true,
              "isProject": false,
              "tags": Array [],
            },
            Object {
              "_id": "item6",
              "isCompleted": false,
              "isProject": false,
              "tags": Array [],
            },
            Object {
              "_id": "item8",
              "isCompleted": true,
              "isProject": false,
              "tags": Array [],
            },
          ],
          "isProject": true,
        },
        null,
        Object {
          "_id": "item7",
          "isCompleted": true,
          "isProject": false,
          "tags": Array [],
        },
      ],
      "isProject": true,
    },
    Object {
      "_id": "item9",
      "isCompleted": false,
      "isProject": false,
      "tags": Array [],
    },
  ],
  "isProject": true,
}
`;
